tests/integration/targets/copr/tasks/main.yml

```diff
@@ -3,14 +3,29 @@
 # GNU General Public License v3.0+ (see LICENSES/GPL-3.0-or-later.txt or https://www.gnu.org/licenses/gpl-3.0.txt)
 # SPDX-License-Identifier: GPL-3.0-or-later
 
-- when: ansible_distribution == 'Fedora'
+- name: Fix improperly set ansible_python_interpreter on Stream 8
+  when:
+    - "ansible_distribution == 'CentOS'"
+    - "ansible_distribution_release == 'Stream'"
+    - "ansible_distribution_major_version == '8'"
+  ansible.builtin.set_fact:
+    ansible_python_interpreter: /usr/libexec/platform-python
```

Comment:
Tests must always use the Python provider configured for CI, and not switch to another one. If that is not possible, tests must be skipped on that platform.

Reply:
The Python interpreter set in CI (/usr/bin/python3.8) is not the default Python interpreter for this platform. The interpreter discovery here is broken. python3.8 is an alternative Python interpreter on EL 8, and it's not supposed to be used to execute Ansible modules. /usr/libexec/platform-python (which is python3.6 installed in a different path due to shenanigans that are irrelevant here) is the default system Python interpreter.

``` console
$ podman run -ti centos:stream8 bash
# dnf install -y ansible-core
# rpm -q ansible-core
ansible-core-2.13.5-1.el8.x86_64
# ansible --version
ansible [core 2.13.5]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/root/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.9/site-packages/ansible
  ansible collection location = /root/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.9.13 (main, Jun 24 2022, 15:32:51) [GCC 8.5.0 20210514 (Red Hat 8.5.0-13)]
  jinja version = 3.1.2
  libyaml = True
# ansible -i localhost, -c local localhost -m setup | grep discovered_interpreter
        "discovered_interpreter_python": "/usr/libexec/platform-python",
```

(this is the same when installing ansible-core with pip. there's no downstream patches here.)

Reply:
There is no interpreter discovery, the interpreter is explicitly set for the image. See https://github.com/ansible-community/images/tree/main/ansible-test.

Reply:
I've submitted https://github.com/ansible-community/images/pull/34.

Reply:
> Ah, I see. In any case, it needs to be fixed. In the image, the necessary Python packages should be installed for python3.6. Collections should change their CI configurations to use python3.6 with that image.

I don't think this needs to be fixed, as testing CentOS 8 Stream with a more modern Python as the system Python is the intended usage of that image.

What we could do is add another image (resp. the same image with another Python version), so that C8S tests also run with only the system Python installed.

Anyway, if this test needs the default Python, it should be skipped if another Python is in use (assuming the module does not respawn).

Reply:
>  testing CentOS 8 Stream with a more modern Python as the system Python is the intended usage of that image.

Ansible modules should almost never be executed with a non-system Python version. This is not supported. Users who file bugs about missing libraries or other issues, whether to me who maintains the ansible community package in EPEL or to Red Hat who maintains ansible-core in RHEL and Stream, get told to use the correct interpreter and the bug is closed as NOTABUG. For modules that rely on system library bindings[^1], this test image is useless. For other modules, it's still wasting time testing alternative setups that nobody uses. Other distributions, such as Fedora, contain multiple alternative Python versions in their repositories that we don't test; I don't see what's different here.

[^1]:  E.g. c.g.copr (dnf and rpm), the firewalld modules in ansible.posix, the selinux module in ansible.posix (libselinux)

> Anyway, if this test needs the default Python, it should be skipped if another Python is in use (assuming the module does not respawn).

I'll skip the test for now.

Reply:
> Ansible modules should almost never be executed with a non-system Python version. This is not supported.

Who says that this is not supported? What's supported or not depends on the module, there's no reason a module cannot support anything but the system Python version.

Reply:
I would say that this setup is **really** common, since most folks that run ansible-core on RHEL/CentOS Stream 8/... **cannot** use the system interpreter since ansible-core requires Python 3.9+. So everyone using ansible-core on these systems **has** to use Python 3.9 or newer to run ansible-core on, so testing plugins and modules with it is the most realistic thing I can think of.

Reply:
The `ansible-core` package uses python3.9, but modules are still supposed to be executed with platform-python/python3.6. There's been multiple upstream changes (revamping interpreter discovery, adding the respawning feature) to make sure that platform-python is used on EL 8.   According to RHEL:

> We recommend to use Python 3.6 or Platform-Python to run Ansible.
> 
> https://www.ansible.com/blog/integrating-ansible-and-red-hat-enterprise-linux-8-beta

-- https://bugzilla.redhat.com/show_bug.cgi?id=1932650#c1

Implicit localhost (which is somewhat common) bypasses interpreter discovery, but besides that, users have to take explicit action to use a different Python interpreter. As long as we're testing python3.6, I suppose it wouldn't be a bad thing to also test python3.9. We'd just have to use this somewhat ugly conditional for modules that don't work with alternative Python stacks.

I guess the situation here does somewhat justify testing alternative Python interpreters on cs8 even though we don't do this for other platforms.

Reply:
We should probably also test this on other platforms that allow it; right now Docker images mostly have one interpreter (except the `default` one), but there are some VMs which have multiple ones: https://github.com/ansible/ansible/blob/devel/test/lib/ansible_test/_data/completion/remote.txt The most interesting ones are probably `freebsd/13.1` and `rhel/8.6`. (And I'm wondering whether we should switch from `ubuntu2204` as a container to `ubuntu2204` as a VM, but that's a separate thing...)

Reply:
If you want to be complete, Fedora 37 has (2.6 and 3.6-3.12) :D.

> The most interesting ones are probably freebsd/13.1 and rhel/8.6

I don't know anything about BSDs, but I don't think it makes sense to test 3.8 or 3.9 on RHEL 8 if we're already doing it for Stream.

> (And I'm wondering whether we should switch from ubuntu2204 as a container to ubuntu2204 as a VM, but that's a separate thing...)

That's probably a good idea. VMs are more representative, and systemd and containers are not the best of friends.